test/fp32-limits.js

```diff
@@ -0,0 +1,78 @@
+import destination from '@turf/destination';
+import {lngLatToWorld, getDistanceScales} from 'viewport-mercator-project';
+import test from 'tape-catch';
+
+import VIEWPORT_PROPS from './utils/sample-viewports';
+
+function getDiff(value, baseValue) {
+  const errorPixels = value.map((v, i) => Math.abs(v - baseValue[i]));
+  const error = value.map(
+    (v, i) => Math.abs(v - baseValue[i]) / Math.min(Math.abs(v), Math.abs(baseValue[i]))
+  );
+
+  return {
+    errorPixels,
+    error,
+    message: `off by \
+      (${errorPixels.map(d => d.toFixed(3)).join(', ')}) pixels, \
+      (${error.map(d => `${(d * 100).toFixed(3)}%`).join(', ')})`
+  };
+}
+
+test('FP32 & Offset Comparison', t => {
+  // Explore limits at different scales
+  for (let zoom = 1; zoom <= 20; zoom++) {
+    const scale = Math.pow(2, zoom);
+    t.comment('--------');
+    t.comment(`Zoom = ${zoom}, Scale = ${scale}`);
+
+    for (const vc in VIEWPORT_PROPS) {
+      t.comment(vc);
+      const props = VIEWPORT_PROPS[vc];
+      const {longitude, latitude} = props;
+      const {pixelsPerMeter, pixelsPerMeter2} = getDistanceScales({
+        latitude,
+        longitude,
+        scale,
+        highPrecision: true
+      });
+
+      // Test distance from one edge of the viewport to the other
+      const delta = Math.pow(2, 20 - zoom) * 50;
+      t.comment(`R = ${delta} meters`);
+
+      // To pixels
+      const offsetPixelPos = [
+        delta * (pixelsPerMeter[0] + pixelsPerMeter2[0] * delta),
+        delta * (pixelsPerMeter[1] + pixelsPerMeter2[1] * delta)
```

Comment:
The experimental fp32 mode is using lnglat offset, not meter offset:

```
const pointFP32 = point.map(f => Math.fround(f));
lngLatToWorld(pointFP32); // classic fp32 shader projection

const point64Low = point.map(f => f - Math.fround(f));
const projectedOffset = [
  point64Low[0] * (pixelsPerDegree[0] + pixelsPerDegree2[0] * point64Low[1]),
  point64Low[1] * (pixelsPerDegree[1] + pixelsPerDegree2[1] * point64Low[1])
];
vec2_add(lngLatToWorld(pointFP32), projectedOffset);  // improved fp32 shader projection

```

Reply:
Just checked your implementation, I was also wrong. Fixed.

Reply:
What did you fix?